compiler/src/locals.rs

```diff
@@ -0,0 +1,43 @@
+use crate::r#type::get_type_stack_size;
+use analyzer::relations::LocalId;
+use analyzer::types::hir::TypeId;
+
+/// contains the different index per local value allocated in the locals area
+pub struct LocalsLayout {
+    /// the start indexes of bound Locals
+    indexes: Vec<Option<u32>>,
+    /// the length in bytes
+    len: u32,
+}
+
+impl LocalsLayout {
+    pub fn fixed_count(count: usize) -> Self {
+        let mut indexes = Vec::with_capacity(count);
+        indexes.resize(count, None);
+        Self { indexes, len: 0 }
+    }
+
+    /// reserves the space in the locals depending on the stack size needed by the given type
+    pub fn initialize_space(&mut self, id: LocalId, ty: TypeId) {
+        if self.indexes[id.0].is_some() {
+            return;
+        }
+        let size: u8 = get_type_stack_size(ty).into();
+        self.indexes[id.0] = Some(self.len);
+        self.len += size as u32;
+    }
```

Comment:
```suggestion
    /// Reserves the space in the locals depending on the stack size needed by the given type.
    ///
    /// Different initialization orders will result in different indexes.
    ///
    /// # Panics
    /// Panics if the local id is out of bounds.
    pub fn initialize_space(&mut self, id: LocalId, stack_size: ValueStackSize) {
        let size: u8 = stack_size.into();
        self.indexes[id.0] = Some(self.total_size);
        self.total_size += size as u32;
    }
```
Do not silently ignore function parameters. Since this struct is responsible for determining the layout, externalize the process of calculating the size of a type.